lib.warn: Use or behave like builtins.warn

roberth opened the pull request:
## Description of changes

Base `lib.warn` on [`builtins.warn`](https://github.com/NixOS/nix/pull/10592) when available.
Behave like it when it's not available.

> A variation of `trace`, but
> 
>   * with the correct log level
> 
>   * no ANSI codes in expressions
> 
>   * can be debugged without stopping for actual `trace` calls (ie intentional "info level" or trace logging).
> 
>   * logs warnings without `trace:` in front of it
> 
>   * follows the Nix color scheme.

Alternative to, and
- Closes https://github.com/NixOS/nixpkgs/pull/306026

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
lib.warn: Use or behave like builtins.warn

lorenzleutgeb commented on lib/trivial.nix:
```diff
@@ -698,9 +701,18 @@ in {
+          # `builtins.warn` requires a string message, so we enforce that in our implementation, so that callers aren't accidentally incompatible with newer Nix versions.
+          assert isString msg;
+          if mustAbort
+          then builtins.trace "[1;31mwarning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
+          else builtins.trace "[1;31mwarning: ${msg}[0m" v
```
```suggestion
          then builtins.trace "[1;31mevaluation warning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
          else builtins.trace "[1;31mevaluation warning: ${msg}[0m" v
```

compare https://github.com/NixOS/nix/pull/10592/files#diff-b3d97fa793934fc7b5faf1ae5e03c8db9fa28cdc4e0a61c47a08785e26805361R243-R244

lorenzleutgeb committed:
lib.warn: Say _evaluation_ warning, like builtins.warn

Aleksanaa commented:
![image](https://github.com/user-attachments/assets/a95d441b-ef17-4f58-83d5-9ba3914581b2)

The color seems not the same?

Aleksanaa commented on lib/trivial.nix:
```diff
@@ -698,9 +701,18 @@ in {
+          assert isString msg;
+          if mustAbort
+          then builtins.trace "[1;31mevaluation warning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
+          else builtins.trace "[1;31mevaluation warning: ${msg}[0m" v
```
```suggestion
          else builtins.trace "[1;35mevaluation warning: ${msg}[0m" v
```

To keep the color the same? But I don’t know why. It seems that Nix doesn’t like the escape sequence in the middle of the output string? So the following msg cannot be output in the default color.

Aleksanaa replied:
Oh, there's a hidden character, my bad

Aleksanaa commented:
After editing it to 35 and moving the hidden character (`\x1b`? Just copied it, forgot how to input it) in front of `${msg}`

Old nix:

![image](https://github.com/user-attachments/assets/d4fc7f6e-71c4-43c0-b34f-5e3fbb978b8a)

New nix:

![image](https://github.com/user-attachments/assets/aeb2cdf5-826a-4cbf-a009-0c8e4738fa1f)

roberth committed:
lib.warn: Fix color

Co-authored-by: Aleksana <alexander.huang.y@gmail.com>

roberth committed:
lib.warn: Remove color from the message itself

This aligns with Nix and makes it more readable.
The prefix still stands out.

roberth commented on lib/trivial.nix:
```diff
@@ -698,9 +701,18 @@ in {
+          # `builtins.warn` requires a string message, so we enforce that in our implementation, so that callers aren't accidentally incompatible with newer Nix versions.
+          assert isString msg;
+          if mustAbort
+          then builtins.trace "[1;31mevaluation warning:[0m ${msg}" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
```
This one's still red, which I think is appropriate because it is fatal.

Aleksanaa replied:
Yeah, fair

infinisil reviewed (changes requested):
Implementation is looking good to me, but the docs of the function need to be updated a bit.

roberth committed:
lib.warn: Update docs

- Bring up to date
- Give meaning to the metavariables
- Use italics for metavariables (just like the Nix manual)
- Don't abbreviate
  - No hard feelings, Val.

roberth commented:
> docs

Done:

   - Bring up to date
   - Give meaning to the metavariables
   - Use italics for metavariables (just like the Nix manual)
   - Don't abbreviate
     - No hard feelings, Val.

infinisil reviewed (approved):
Looking good!

roberth commented:
Thanks!

infinisil commented:
Weird to see the "Delete branch" button on somebody else's PR, instinctively clicked it, not sure if you want to keep it around :P
